### ISSUE TITLE
colexec: suboptimal behavior of unordered distinct with limit

### ISSUE THREAD

**Issue by xinyuliu12** [cockroachdb/cockroach#57566]
Hello everyone!

First off, I am sorry for gestures vaguely all of this. Second, I think we may have found a performance bug related to projecting a column within the filter.

Here is a pair of TPC-H queries that exhibit this behavior.

```
SELECT   "l_linenumber", 
         "l_shipinstruct", 
         "l_suppkey" AS "c_2", 
         "l_returnflag", 
         "l_suppkey", 
         "l_discount" 
FROM     "lineitem" 
WHERE    "l_returnflag" = 'N' 
GROUP BY "l_linenumber", 
         "l_shipinstruct", 
         "l_suppkey", 
         "l_returnflag", 
         "l_discount"FETCH next 13 rows only;


SELECT   "l_linenumber", 
         "l_shipinstruct", 
         "l_suppkey"          AS "c_2", 
         Cast('N' AS CHAR(1)) AS "l_returnflag", 
         "l_suppkey", 
         "l_discount" 
FROM     "lineitem" 
WHERE    "l_returnflag" = 'N' 
GROUP BY "l_suppkey", 
         "l_linenumber", 
         "l_discount", 
         "l_shipinstruct"FETCH next 13 rows only;


```

**Actual Behavior**
The first query takes 9 seconds on the v20.2.0-alpha.1, while the second query only takes 53 millisecond.
The first query's "EXPLAIN ANALYZE (DEBUG)" result is 
[stmt-bundle-firstquery.zip](https://github.com/cockroachdb/cockroach/files/5643841/stmt-bundle-firstquery.zip)

The second query's "EXPLAIN ANALYZE (DEBUG)" result is 
[stmt-bundle-secondquery.zip](https://github.com/cockroachdb/cockroach/files/5643842/stmt-bundle-secondquery.zip)

**Expected Behavior**
I would have expected the database system run these two queries with similar execution time, given that they return the same result tables (the second query projects `l_returnflag` as a literal column). 
Here are the steps for reproducing our observations:

**Test environment**

* Ubuntu 20.04 machine "Linux panda 5.4.0-40-generic #44-Ubuntu SMP Tue Jun 23 00:01:04 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux"
* Cockroachdb v20.2.0-alpha.1
* Database: TPC-H benchmark (with scale factor 5)

**Reproduce Bug**

1. Install cockroachdb v20.2.0-alpha.1     

```
    $ mkdir -p $(go env GOPATH)/src/github.com/cockroachdb
    $ cd $(go env GOPATH)/src/github.com/cockroachdb
    $ git clone https://github.com/cockroachdb/cockroach
    $ cd cockroach
    $ git checkout eaa939c
    $ make build
    $ sudo make install
```

* start a cockroachdb node in your preferred working directory
```
    $ cockroach start --insecure --store=node1 --listen-addr=localhost:26257 --http-addr=localhost:8080 --join=localhost:26257,localhost:26258,localhost:26259 --background
    $ cockroach init --insecure --host=localhost:26257
```

2. Set up TPC-H test benchmark (if you already have a TPC-H benchmark set up, you can skip step2; starting from step2, all commands should run in the directory ./tpch5 after downloading and extracting it)

* Download TPC-H (scale factor of 5) and extract it
    Download the dataset from the link: https://drive.google.com/file/d/1MLVFPKdRQhlRgmeRIumyIj8NBMRSQ3um/view?usp=sharing
```
    $ tar xzvf tpch5_setup.tar.gz
    $ cd tpch5
    $ chmod +x setup.sh
```

* Create DB and TPC-H schema
`   $ cockroach sql --insecure --host=localhost:26257`    
    In CockroachDB's built-in SQL client to create database:
```
    $ create database tpch5;
    $ exit
```
* Create TPC-H Schema
   `    $ cockroach sql --insecure --host=localhost:26257 -d tpch5 < dss.ddl
`
* Set up benchmark
```  
    $ ./setup.sh
    $ cockroach sql --insecure --host=localhost:26257 -d tpch5 < dss.ri

```
3. Test SQL query that exhibits performance issue
* Execute the queries after logging to the testing benchmark using the following command
`    $ cockroach sql --insecure --host=localhost:26257 -d tpch5
`

**Comment by yuzefovich** [comment]
Note that the second query doesn't group by `l_returnflag`, not sure if it matters.

cc @RaduBerinde

**Comment by yuzefovich** [comment]
Oh, didn't realize there were bundles attached.

The vectorized engine is not used for the second query because of the observed type mismatch during the vectorization:
```
 9223372036854.777ms      0.025ms    event:sql/colexec/colbuilder/execplan.go:1197 [n1,client=127.0.0.1:57660,hostnossl,user=root] planning render expressions [@2 @4 @1 'N':::STRING @1 @3]
 9223372036854.777ms      0.097ms    event:sql/distsql_running.go:203 [n1,client=127.0.0.1:57660,hostnossl,user=root] failed to vectorize: mismatched types at index 3: expected [int char int char int decimal]	actual [int char int string int decimal] 
```
This has already been fixed on master, but on 20.2 version in such cases we fallback to the row engine.

Hm, it's an interesting case that highlights the current limitations of the vectorized engine - we don't propagate the limit to other operators, so they aren't aware of it, and we also buffer up all distinct tuples in `unordered distinct` before emitting any output. This is definitely an execution problem.

**Comment by yuzefovich** [comment]
I think the vectorized engine has a problem with limit propagation only in case of the buffering operators - hash aggregation, hash join, unordered distinct; for all other operators the limit propagation should just work.

I've opened up a PR to address the unordered distinct inefficiency. Hash aggregation cannot be optimized because we do need to consume the whole input before we can emit any output.

Hash join is the only remaining case where the inefficiency will remain. I'll open up a separate issue to track that since addressing that is not trivial.
